brainiak/matnormal/utils.py

```diff
@@ -0,0 +1,124 @@
+import tensorflow as tf
+import tensorflow_probability as tfp
+from scipy.stats import norm
+from numpy.linalg import cholesky
+import numpy as np
+
+
+def rmn(rowcov, colcov):
+    """
+    Generate random draws from a zero-mean matrix-normal distribution.
+
+    Parameters
+    -----------
+    rowcov : np.ndarray
+        Row covariance (assumed to be positive definite)
+    colcov : np.ndarray
+        Column covariance (assumed to be positive definite)
+    """
+
+    Z = norm.rvs(size=(rowcov.shape[0], colcov.shape[0]))
+    return cholesky(rowcov).dot(Z).dot(cholesky(colcov))
+
+
+def xx_t(x):
+    """
+    Outer product
+    :math:`xx^T`
+
+    Parameters
+    -----------
+    x : tf.Tensor
+
+    """
+    return tf.matmul(x, x, transpose_b=True)
+
+
+def x_tx(x):
+    """Inner product
+    :math:`x^T x`
+
+    Parameters
+    -----------
+    x : tf.Tensor
+
+    """
+    return tf.matmul(x, x, transpose_a=True)
+
+
+def scaled_I(x, size):
+    """Scaled identity matrix
+    :math:`x I_{size}`
+
+    Parameters
+    ------------
+    x: float or coercable to float
+        Scale to multiply the identity matrix by
+    size: int or otherwise coercable to a size
+        Dimension of the scaled identity matrix to return
+    """
+    return tf.linalg.tensor_diag(tf.ones([size], dtype=tf.float64) * x)
+
+
+def flatten_cholesky_unique(L):
+    """
+    Flattens nonzero-elements Cholesky (triangular) factor
+    into a vector, and logs diagonal to make parameterization
+    unique. Inverse of unflatten_cholesky_unique.
+    """
+    L[np.diag_indices_from(L)] = np.log(np.diag(L))
```

Comment:
This change is in-place. I wonder if there is a risk of unintentionally change variable (L) outside the function here? Is it better to make a copy of L before changing the diagonal?

Reply:
I ended up replacing the numpy calls with TF calls for consistency. The new call (https://www.tensorflow.org/api_docs/python/tf/linalg/set_diag) returns a tensor the same as the input tensor but with a changed diagonal, i.e. it is a copy.